src/game_map.cpp

```diff
@@ -918,25 +935,61 @@ void Game_Map::UpdateEncounterSteps() {
 		return;
 	}
 
-	location.encounter_steps = location.encounter_steps - terrain->encounter_rate;
+	location.encounter_steps += terrain->encounter_rate;
 
-	if (location.encounter_steps <= 0) {
-		ResetEncounterSteps();
-		PrepareEncounter();
+	struct Row {
+		int ratio;
+		float pmod;
+	};
+
+#if 1
+	static constexpr Row enc_table[] = {
+		{ 0, 0.0625},
+		{ 20, 0.125 },
+		{ 40, 0.25 },
+		{ 60, 0.5 },
+		{ 100, 2.0 },
+		{ 140, 4.0 },
+		{ 160, 8.0 },
+		{ 180, 16.0 },
+		{ INT_MAX, 16.0 }
+	};
+#else
+	//Old versions of RM2k used this table.
+	//Left here for posterity.
+	static constexpr Row enc_table[] = {
+		{ 0, 0.5 },
+		{ 20, 2.0 / 3.0 },
+		{ 50, 5.0 / 6.0 },
+		{ 100, 6.0 / 5.0 },
+		{ 200, 3.0 / 2.0 },
+		{ INT_MAX, 3.0 / 2.0 }
+	};
+#endif
+	const auto encounter_rate = GetEncounterRate();
+	const auto ratio = location.encounter_steps / encounter_rate;
+
+	auto& idx = last_encounter_idx;
+	while (ratio > enc_table[idx+1].ratio) {
+		++idx;
 	}
-}
+	const auto& row = enc_table[idx];
 
-void Game_Map::ResetEncounterSteps() {
-	int rate = GetEncounterRate();
-	if (rate > 0) {
-		int throw_one = Utils::GetRandomNumber(0, rate - 1);
-		int throw_two = Utils::GetRandomNumber(0, rate - 1);
+	const auto pmod = row.pmod;
+	const auto p = (1.0f / float(encounter_rate)) * pmod * (float(terrain->encounter_rate) / 100.0f);
+	auto draw = std::uniform_real_distribution<float>()(Utils::GetRNG());
```

Comment:
is std::uniform_real_distribution implementation standardized? I mean does it give the same values for the same seed independent of the C++ implementation (clang c++ lib, gcc c++ lib, MSVC)?

Reply:
Good question. I can't find anything that guarantees this. 

`uniform_real_distribution` does not actually play a part in drawing random numbers. The rng stored in `Utils::GetRNG()` does that. The distribution is just a deterministic mathematical function which takes an input random number in some integer range and converts it to floating point. I guess standard library implementations could use different combinations of floating point expressions which could result is slightly different answers.

That being said, this is the right way to do this. Modulus and all the other tricks we used in the past with C `rand()` are bad. They don't actually give you good random numbers.

https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution

Reply:
Also, Here is a great cppcon talk on the C++11 random facilities. Worth the time to watch.

https://www.youtube.com/watch?v=6DPkyvkMkk8

Reply:
I will change this to use `PercentChance()` when #1487 gets merged.